### PR TITLE
Resolve all signatures before fetching functions

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-geospatial/src/test/java/io/prestosql/plugin/geospatial/TestSpatialPartitioningInternalAggregation.java [hunk 1]
```diff
@@ -96,12 +96,10 @@ public void test(int partitionCount)
...
-        return functionAssertions.getMetadata().getAggregateFunctionImplementation(new Signature(
-                "spatial_partitioning",
-                FunctionKind.AGGREGATE,
-                VARCHAR.getTypeSignature(),
-                GEOMETRY.getTypeSignature(),
-                INTEGER.getTypeSignature()));
+        Metadata metadata = functionAssertions.getMetadata();
+        return metadata.getAggregateFunctionImplementation(metadata.resolveFunction(
+                QualifiedName.of("spatial_partitioning"),
+                fromTypeSignatures(GEOMETRY.getTypeSignature(), INTEGER.getTypeSignature())));
```
Comment (martint) [inline presto-geospatial/src/test/java/io/prestosql/plugin/geospatial/TestSpatialPartitioningInternalAggregation.java]: `fromTypes`

File: presto-geospatial/src/test/java/io/prestosql/plugin/geospatial/aggregation/AbstractTestGeoAggregationFunctions.java [hunk 2]
```diff
@@ -48,12 +49,11 @@ public void registerFunctions()
...
-        function = functionAssertions.getMetadata().getAggregateFunctionImplementation(new Signature(
-                getFunctionName(),
-                FunctionKind.AGGREGATE,
-                GEOMETRY.getTypeSignature(),
-                GEOMETRY.getTypeSignature()));
+        Metadata metadata = functionAssertions.getMetadata();
+        metadata.addFunctions(extractFunctions(plugin.getFunctions()));
+        function = metadata.getAggregateFunctionImplementation(metadata.resolveFunction(
+                QualifiedName.of(getFunctionName()),
+                fromTypeSignatures(GEOMETRY.getTypeSignature())));
```
Comment (martint) [inline presto-geospatial/src/test/java/io/prestosql/plugin/geospatial/aggregation/AbstractTestGeoAggregationFunctions.java]: `fromTypes`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -1504,7 +1504,7 @@ private void analyzeWindowFunctions(QuerySpecification node, List<Expression> ou
 
                 List<TypeSignature> argumentTypes = mappedCopy(windowFunction.getArguments(), expression -> analysis.getType(expression).getTypeSignature());
 
-                FunctionKind kind = metadata.resolveFunction(windowFunction.getName(), fromTypeSignatures(argumentTypes)).getKind();
+                FunctionKind kind = metadata.resolveFunction(windowFunction.getName(), fromTypeSignatures(argumentTypes)).getSignature().getKind();
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: `fromTypes`

File: presto-main/src/main/java/io/prestosql/cost/StatsUtil.java [hunk 4]
```diff
@@ -40,8 +40,8 @@ static OptionalDouble toStatsRepresentation(Metadata metadata, Session session,
     {
         if (convertibleToDoubleWithCast(type)) {
             InterpretedFunctionInvoker functionInvoker = new InterpretedFunctionInvoker(metadata);
-            Signature castSignature = metadata.getCoercion(type, DoubleType.DOUBLE);
-            return OptionalDouble.of((double) functionInvoker.invoke(castSignature, session.toConnectorSession(), singletonList(value)));
+            ResolvedFunction castFunction = metadata.getCoercion(type, DoubleType.DOUBLE);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/cost/StatsUtil.java]: static import

File: presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java [hunk 5]
```diff
@@ -96,69 +98,89 @@ else if (keyType.getJavaType() == Slice.class) {
...
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final ResolvedFunction castFunction;
+
+        public MissingKeyExceptionFactory(Metadata metadata, Type keyType)
+        {
+            functionInvoker = new InterpretedFunctionInvoker(metadata);
+
+            ResolvedFunction castFunction = null;
+            try {
+                castFunction = metadata.getCoercion(keyType, VarcharType.VARCHAR);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java]: static import

File: presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java [hunk 6]
```diff
@@ -96,69 +98,89 @@ else if (keyType.getJavaType() == Slice.class) {
...
     {
-        String stringValue;
-        try {
-            stringValue = ((Slice) functionInvoker.invoke(castSignature(VarcharType.VARCHAR, type), session, value)).toStringUtf8();
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final ResolvedFunction castFunction;
+
+        public MissingKeyExceptionFactory(Metadata metadata, Type keyType)
+        {
+            functionInvoker = new InterpretedFunctionInvoker(metadata);
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java]: That actually makes the code to construct of this private internal class much more complex.  We could use a static factory method, but that would be the same as the constructor.

File: presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java [hunk 7]
```diff
@@ -665,7 +663,15 @@ protected RowExpression visitIsNullPredicate(IsNullPredicate node, Void context)
...
         protected RowExpression visitNotExpression(NotExpression node, Void context)
         {
-            return call(Signatures.notSignature(), BOOLEAN, process(node.getValue(), context));
+            return notExpression(process(node.getValue(), context));
+        }
+
+        private RowExpression notExpression(RowExpression value)
+        {
+            return new CallExpression(
+                    metadata.resolveFunction(QualifiedName.of("not"), fromTypes(BOOLEAN)),
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java]: I eventually remove that class entirely, so it I'd prefer not to move this in for now.

File: presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java [hunk 8]
```diff
@@ -450,15 +444,22 @@ protected RowExpression visitCast(Cast node, Void context)
...
+            Type returnType = getType(node);
             if (node.isTypeOnly()) {
-                return changeType(value, getType(node));
+                return changeType(value, returnType);
             }
 
             if (node.isSafe()) {
-                return call(tryCastSignature(getType(node), value.getType()), getType(node), value);
+                return call(
+                        metadata.getCoercion(QualifiedName.of("TRY_CAST"), value.getType().getTypeSignature(), returnType.getTypeSignature()),
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java]: Same

File: presto-main/src/test/java/io/prestosql/sql/planner/TestSortExpressionExtractor.java [hunk 9]
```diff
@@ -113,4 +139,38 @@ private static void assertGetSortExpression(Expression expression, String expect
         Optional<SortExpressionContext> actual = SortExpressionExtractor.extractSortExpression(BUILD_SYMBOLS, expression);
         assertEquals(actual, expected);
     }
+
+    private static Expression resolveFunctions(Metadata metadata, Session session, TypeProvider typeProvider, Expression expression)
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestSortExpressionExtractor.java]: Updated to `ExpressionTestUtils.planExpression`

File: presto-main/src/main/java/io/prestosql/cost/StatsUtil.java (hunk 4 above)
Comment (dain) [reply presto-main/src/main/java/io/prestosql/cost/StatsUtil.java]: I added a commit at the end to clean up this class.

**Pull request merged**